elk/run.py

```diff
@@ -88,28 +90,36 @@ def prepare_data(
 
         return x0, x1, val_x0, val_x1, train_labels, val_labels
 
-    def apply_to_layers(self, func):
-        """Apply a function to each layer of the dataset in parallel."""
-
-        devices = select_usable_devices(self.cfg.num_gpus)
-        num_devices = len(devices)
+    def apply_to_layers(
+        self,
+        func: Callable[[int], Log],
+        num_devices: int,
+        to_csv_line: Callable[[Log], list[str]],
+        csv_columns: list[str],
+    ):
```

Comment:
i made func have a signature like
```
Callable[[int], Log]
```

previously it was something like

```
Callable[[int, int, list[int]], Log] # layer, device, list of devices
```

which could be pretty confusing